usgpo/get-links.py

```diff
@@ -0,0 +1,187 @@
+import argparse
+import json
+import os
+import queue
+import time
+from concurrent.futures import ThreadPoolExecutor, as_completed
+
+import jsonlines
+import requests
+from tqdm.auto import tqdm
+
+from licensed_pile import logs
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--api-key", required=True, help="GovInfo API key")
+    parser.add_argument(
+        "--start-date",
+        required=True,
+        help="Start date in ISO8601 format (yyyy-MM-dd'T'HH:mm:ss'Z')",
+    )
+    parser.add_argument("--output-dir", required=True, help="Path to output directory")
+    parser.add_argument("--workers", type=int, default=20, help="Number of threads")
+    parser.add_argument(
+        "--collections",
+        nargs="+",
+        default=[
+            "BILLS",
+            "BUDGET",
+            "CDIR",
+            "CFR",
+            "CPD",
+            "CRI",
+            "CZIC",
+            "GAOREPORTS",
+            "GOVPUB",
+            "GPO",
+            "HJOURNAL",
+            "HOB",
+            "PAI",
+            "PLAW",
+            "USCODE",
+        ],
+    )
+    args = parser.parse_args()
+    return args
+
+
+def api_query(endpoint, headers, params):
+    logger = logs.get_logger("usgpo")
+    response = requests.get(endpoint, headers=headers, params=params)
+    if response.status_code == 429:
+        # Sleep for an hour if we've hit the rate-limit
+        logger.info("Sleeping for one hour to avoid rate-limit")
+        time.sleep(60 * 60)
+        response = requests.get(endpoint, headers=headers, params=params)
+    return response
+
+
+def get_collections(api_key):
+    logger = logs.get_logger("usgpo")
+    response = api_query(
+        "https://api.govinfo.gov/collections",
+        headers={"accept": "application/json"},
+        params={"api_key": args.api_key},
+    )
+    if response.status_code == 200:
+        output = response.json()
+        for record in output["collections"]:
+            yield record["collectionCode"]
+    else:
+        logger.error(f"get_collections received status code {response.status_code}")
+
+
+def get_packages(api_key, collections, start_date, package_queue):
+    logger = logs.get_logger("usgpo")
+    url = f"https://api.govinfo.gov/published/{start_date}"
+    offset_mark = "*"
+    pbar = tqdm(desc="Producer")
+    while url is not None:
+        response = api_query(
+            url,
+            headers={"accept": "application/json"},
+            params={
+                "api_key": args.api_key,
+                "offsetMark": offset_mark,
+                "pageSize": 1000,
+                "collection": ",".join(collections),
+            },
+        )
+        if response.status_code == 200:
+            output = response.json()
+
+            for record in output["packages"]:
+                package_queue.put(record)
+                pbar.update(1)
+
+            url = output["nextPage"]
+            offset_mark = None
```

Comment:
Can you document this `offset_mark` a bit, being `"*"` the first time and `None` the rest is a bit confusing.